# 01. Importing Libraries

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

# 02. Importinig dataframe

In [2]:
# set file path
path = r'C:\Users\campa\data projects\instacart basket project'

In [3]:
#import dataframe
ords_prods_merged = pd.read_pickle(os.path.join(path, '02 data', 'prepared data', 'orders_products_derived.pkl'))

In [ ]:
#checking shape
ords_prods_merged.shape

In [5]:
ords_prods_merged.head()

Unnamed: 0_x  order_id  user_id  order_number  order_day_of_week  \
0             0   2539329        1             1                  2   
1             1   2398795        1             2                  3   
2             2    473747        1             3                  3   
3             3   2254736        1             4                  4   
4             4    431534        1             5                  4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  7                   15.0         196                  1   
2                 12                   21.0         196                  1   
3                  7                   29.0         196                  1   
4                 15                   28.0         196                  1   

   reordered  Unnamed: 0_y product_name  aisle_id  department_id  prices  \
0          0           195         Soda        77              7     9.0   
1          1           195         Soda        77              7     9.0   
2          1           195         Soda        77              7     9.0   
3          1           195         Soda        77              7     9.0   
4          1           195         Soda        77              7     9.0   

     price_range_loc    Busiest Days busiest_period_of_day  
0  Mid-Range Product  Regularly Busy        Average Orders  
1  Mid-Range Product      Least Busy        Average Orders  
2  Mid-Range Product      Least Busy           Most Orders  
3  Mid-Range Product      Least Busy        Average Orders  
4  Mid-Range Product      Least Busy           Most Orders

In [ ]:
ords_prods_merged = ords_prods_merged.drop(columns = ['Unnamed: 0_y'])

In [4]:
ords_prods_merged.head()

Unnamed: 0_x  order_id  user_id  order_number  order_day_of_week  \
0             0   2539329        1             1                  2   
1             1   2398795        1             2                  3   
2             2    473747        1             3                  3   
3             3   2254736        1             4                  4   
4             4    431534        1             5                  4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  7                   15.0         196                  1   
2                 12                   21.0         196                  1   
3                  7                   29.0         196                  1   
4                 15                   28.0         196                  1   

   reordered  Unnamed: 0_y product_name  aisle_id  department_id  prices  \
0          0           195         Soda        77              7     9.0   
1          1           195         Soda        77              7     9.0   
2          1           195         Soda        77              7     9.0   
3          1           195         Soda        77              7     9.0   
4          1           195         Soda        77              7     9.0   

     price_range_loc    Busiest Days busiest_period_of_day  
0  Mid-Range Product  Regularly Busy        Average Orders  
1  Mid-Range Product      Least Busy        Average Orders  
2  Mid-Range Product      Least Busy           Most Orders  
3  Mid-Range Product      Least Busy        Average Orders  
4  Mid-Range Product      Least Busy           Most Orders

In [5]:
ords_prods_merged.shape

(32404859, 18)

# 03. Aggregating data with transform

In [6]:
#deriving new variable for maximum amount of orders for each user
ords_prods_merged['max_order'] = ords_prods_merged.groupby(['user_id'])['order_number'].transform(np.max)

In [7]:
ords_prods_merged.head(15)

Unnamed: 0_x  order_id  user_id  order_number  order_day_of_week  \
0              0   2539329        1             1                  2   
1              1   2398795        1             2                  3   
2              2    473747        1             3                  3   
3              3   2254736        1             4                  4   
4              4    431534        1             5                  4   
5              5   3367565        1             6                  2   
6              6    550135        1             7                  1   
7              7   3108588        1             8                  1   
8              8   2295261        1             9                  1   
9              9   2550362        1            10                  4   
10           144   2968173       15            15                  1   
11           146   1870022       15            17                  2   
12           147   1911383       15            18                  2   
13           150   2715276       15            21                  1   
14           151    487368       15            22                  1   

    order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                   8                    NaN         196                  1   
1                   7                   15.0         196                  1   
2                  12                   21.0         196                  1   
3                   7                   29.0         196                  1   
4                  15                   28.0         196                  1   
5                   7                   19.0         196                  1   
6                   9                   20.0         196                  1   
7                  14                   14.0         196                  2   
8                  16                    0.0         196                  4   
9                   8                   30.0         196                  1   
10                  9                    7.0         196                  2   
11                 16                    8.0         196                  6   
12                 11                    7.0         196                  1   
13                  9                    7.0         196                  1   
14                 10                   14.0         196                  1   

    reordered  Unnamed: 0_y product_name  aisle_id  department_id  prices  \
0           0           195         Soda        77              7     9.0   
1           1           195         Soda        77              7     9.0   
2           1           195         Soda        77              7     9.0   
3           1           195         Soda        77              7     9.0   
4           1           195         Soda        77              7     9.0   
5           1           195         Soda        77              7     9.0   
6           1           195         Soda        77              7     9.0   
7           1           195         Soda        77              7     9.0   
8           1           195         Soda        77              7     9.0   
9           1           195         Soda        77              7     9.0   
10          0           195         Soda        77              7     9.0   
11          1           195         Soda        77              7     9.0   
12          1           195         Soda        77              7     9.0   
13          1           195         Soda        77              7     9.0   
14          1           195         Soda        77              7     9.0   

      price_range_loc    Busiest Days busiest_period_of_day  max_order  
0   Mid-Range Product  Regularly Busy        Average Orders         10  
1   Mid-Range Product      Least Busy        Average Orders         10  
2   Mid-Range Product      Least Busy           Most Orders         10  
3   Mid-Range Product      Least Busy        Average Orders  

# 04. Creating Loyalty Flag

In [8]:
#creating flag for loyalty customers
ords_prods_merged.loc[ords_prods_merged['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merged.loc[(ords_prods_merged['max_order'] <= 40) & (ords_prods_merged['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merged.loc[ords_prods_merged['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [ ]:
ords_prods_merged['loyalty_flag'].value_counts()

In [9]:
#checking loyalty flag stats
ords_prods_merged[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3

# 05. Find Spending Habits of Loyalty Levels

In [11]:
#finding average prices for users to compare loyalty levels
ords_prods_merged['Average_spending']= ords_prods_merged.groupby(['user_id'])['prices'].transform(np.mean)


In [22]:
ords_prods_merged.head()

Unnamed: 0_x  order_id  user_id  order_number  order_day_of_week  \
0             0   2539329        1             1                  2   
1             1   2398795        1             2                  3   
2             2    473747        1             3                  3   
3             3   2254736        1             4                  4   
4             4    431534        1             5                  4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  7                   15.0         196                  1   
2                 12                   21.0         196                  1   
3                  7                   29.0         196                  1   
4                 15                   28.0         196                  1   

   reordered product_name  aisle_id  department_id  prices    price_range_loc  \
0          0         Soda        77              7     9.0  Mid-Range Product   
1          1         Soda        77              7     9.0  Mid-Range Product   
2          1         Soda        77              7     9.0  Mid-Range Product   
3          1         Soda        77              7     9.0  Mid-Range Product   
4          1         Soda        77              7     9.0  Mid-Range Product   

     Busiest Days busiest_period_of_day  max_order  loyalty_flag  \
0  Regularly Busy        Average Orders         10  New customer   
1      Least Busy        Average Orders         10  New customer   
2      Least Busy           Most Orders         10  New customer   
3      Least Busy        Average Orders         10  New customer   
4      Least Busy           Most Orders         10  New customer   

   Average_spending  
0          13.29467  
1          13.29467  
2          13.29467  
3          13.29467  
4          13.29467

In [12]:
#checking stats for Average_spending column
ords_prods_merged[['user_id', 'loyalty_flag', 'Average_spending']].head(60)

user_id      loyalty_flag  Average_spending
0         1      New customer          6.367797
1         1      New customer          6.367797
2         1      New customer          6.367797
3         1      New customer          6.367797
4         1      New customer          6.367797
5         1      New customer          6.367797
6         1      New customer          6.367797
7         1      New customer          6.367797
8         1      New customer          6.367797
9         1      New customer          6.367797
10       15  Regular customer          3.980556
11       15  Regular customer          3.980556
12       15  Regular customer          3.980556
13       15  Regular customer          3.980556
14       15  Regular customer          3.980556
15       19      New customer          8.372059
16       19      New customer          8.372059
17       19      New customer          8.372059
18       21  Regular customer          7.012683
19       31  Regular customer          8.137793
20       31  Regular customer          8.137793
21       43  Regular customer          7.444898
22       43  Regular customer          7.444898
23       52  Regular customer          6.356287
24       52  Regular customer          6.356287
25       52  Regular customer          6.356287
26       52  Regular customer          6.356287
27       52  Regular customer          6.356287
28       52  Regular customer          6.356287
29       52  Regular customer          6.356287
30       52  Regular customer          6.356287
31       52  Regular customer          6.356287
32       52  Regular customer          6.356287
33       52  Regular customer          6.356287
34       52  Regular customer          6.356287
35       52  Regular customer          6.356287
36       52  Regular customer          6.356287
37       67  Regular customer          7.656790
38       67  Regular customer          7.656790
39       67  Regular customer          7.656790
40       67  Regular customer          7.656790
41       67  Regular customer          7.656790
42       67  Regular customer          7.656790
43       67  Regular customer          7.656790
44       67  Regular customer          7.656790
45       67  Regular customer          7.656790
46       67  Regular customer          7.656790
47       67  Regular customer          7.656790
48       67  Regular customer          7.656790
49       67  Regular customer          7.656790
50       67  Regular customer          7.656790
51       67  Regular customer          7.656790
52       67  Regular customer          7.656790
53       67  Regular customer          7.656790
54       67  Regular customer          7.656790
55       67  Regular customer          7.656790
56       81      New customer          8.165385
57       81      New customer          8.165385
58       82  Regular customer          6.441221
59       82  Regular customer          6.441221

# 06. Creating Spending Flag

In [13]:
#Creating Spending Flag
ords_prods_merged.loc[ords_prods_merged['Average_spending'] < 10, 'Spending_Flag'] = 'Low Spender'
ords_prods_merged.loc[ords_prods_merged['Average_spending'] >= 10, 'Spending_Flag'] = 'High Spender' 

In [14]:
ords_prods_merged.head()

Unnamed: 0_x  order_id  user_id  order_number  order_day_of_week  \
0             0   2539329        1             1                  2   
1             1   2398795        1             2                  3   
2             2    473747        1             3                  3   
3             3   2254736        1             4                  4   
4             4    431534        1             5                  4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  7                   15.0         196                  1   
2                 12                   21.0         196                  1   
3                  7                   29.0         196                  1   
4                 15                   28.0         196                  1   

   reordered  ...  aisle_id department_id  prices    price_range_loc  \
0          0  ...        77             7     9.0  Mid-Range Product   
1          1  ...        77             7     9.0  Mid-Range Product   
2          1  ...        77             7     9.0  Mid-Range Product   
3          1  ...        77             7     9.0  Mid-Range Product   
4          1  ...        77             7     9.0  Mid-Range Product   

     Busiest Days busiest_period_of_day max_order  loyalty_flag  \
0  Regularly Busy        Average Orders        10  New customer   
1      Least Busy        Average Orders        10  New customer   
2      Least Busy           Most Orders        10  New customer   
3      Least Busy        Average Orders        10  New customer   
4      Least Busy           Most Orders        10  New customer   

   Average_spending Spending_Flag  
0          6.367797   Low Spender  
1          6.367797   Low Spender  
2          6.367797   Low Spender  
3          6.367797   Low Spender  
4          6.367797   Low Spender  

[5 rows x 22 columns]

In [15]:
ords_prods_merged['Spending_Flag'].value_counts()

Low Spender     31770614
High Spender      634245
Name: Spending_Flag, dtype: int64

# 07. Creating Customer Frequency Flag

In [32]:
#deriving variable for median days since last order
ords_prods_merged['median_days_between_orders'] = ords_prods_merged.groupby(['user_id'])['days_since_last_order'].transform(np.median)

In [34]:
ords_prods_merged[['user_id', 'median_days_between_orders']].head(50)

user_id  median_days_between_orders
0         1                        20.5
1         1                        20.5
2         1                        20.5
3         1                        20.5
4         1                        20.5
5         1                        20.5
6         1                        20.5
7         1                        20.5
8         1                        20.5
9         1                        20.5
10       15                        10.0
11       15                        10.0
12       15                        10.0
13       15                        10.0
14       15                        10.0
15       19                         6.0
16       19                         6.0
17       19                         6.0
18       21                         7.5
19       31                         6.0
20       31                         6.0
21       43                        10.0
22       43                        10.0
23       52                         8.0
24       52                         8.0
25       52                         8.0
26       52                         8.0
27       52                         8.0
28       52                         8.0
29       52                         8.0
30       52                         8.0
31       52                         8.0
32       52                         8.0
33       52                         8.0
34       52                         8.0
35       52                         8.0
36       52                         8.0
37       67                         7.0
38       67                         7.0
39       67                         7.0
40       67                         7.0
41       67                         7.0
42       67                         7.0
43       67                         7.0
44       67                         7.0
45       67                         7.0
46       67                         7.0
47       67                         7.0
48       67                         7.0
49       67                         7.0

In [36]:
#Creating Order_Frequency Flag
ords_prods_merged.loc[ords_prods_merged['median_days_between_orders'] > 20, 'Order_Frequency'] = 'Non-Frequent customer'
ords_prods_merged.loc[(ords_prods_merged['median_days_between_orders'] <= 20) &(ords_prods_merged['median_days_between_orders'] > 10), 'Order_Frequency'] = 'Regular Customer'
ords_prods_merged.loc[ords_prods_merged['median_days_between_orders'] <= 10, 'Order_Frequency'] = 'Frequent customer'

In [38]:
#Checking Frequency Flag Stats
ords_prods_merged[['user_id', 'Order_Frequency']].head(20)

user_id        Order_Frequency
0         1  Non-Frequent customer
1         1  Non-Frequent customer
2         1  Non-Frequent customer
3         1  Non-Frequent customer
4         1  Non-Frequent customer
5         1  Non-Frequent customer
6         1  Non-Frequent customer
7         1  Non-Frequent customer
8         1  Non-Frequent customer
9         1  Non-Frequent customer
10       15      Frequent customer
11       15      Frequent customer
12       15      Frequent customer
13       15      Frequent customer
14       15      Frequent customer
15       19      Frequent customer
16       19      Frequent customer
17       19      Frequent customer
18       21      Frequent customer
19       31      Frequent customer

In [39]:
ords_prods_merged['Order_Frequency'].value_counts()

Frequent customer        21559853
Regular Customer          7208564
Non-Frequent customer     3636437
Name: Order_Frequency, dtype: int64

# 08. Comparing Subset Stats to Full DF Stats

In [40]:
#creating subset
df = ords_prods_merged[:1000000]

In [41]:
#checking department order stats of subset
df.groupby('department_id').agg({'order_number' : ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

In [42]:
#checking department order stats of full df
ords_prods_merged.groupby('department_id').agg({'order_number' : ['mean', 'min', max]})

order_number        
                      mean min max
department_id                     
1                15.457838   1  99
2                17.277920   1  99
3                17.170395   1  99
4                17.811403   1  99
5                15.215751   1  99
6                16.439806   1  99
7                17.225802   1  99
8                15.340650   1  99
9                15.895474   1  99
10               20.197148   1  99
11               16.170638   1  99
12               15.887671   1  99
13               16.583536   1  99
14               16.773669   1  99
15               16.165037   1  99
16               17.665606   1  99
17               15.694469   1  99
18               19.310397   1  99
19               17.177343   1  99
20               16.473447   1  99
21               22.902379   1  99

The min and max of the full set and the subset are the same because the minimum is always 1 for this data set and python caps the output for max at 99 as a default.  The main diffences are that there is a row for every department because in the full set every department has sales data and the means are different taking into account the full data set.  Some of the means are quite different eg.dept 17 which is ~15.7 int he subset vs ~ 11.3 in the full set but some are almost identical eg dept 7 which is 17.4 in the subset and 17.2 in the main set.  For mins and maxes the difference between the subset and full set are inconsiquential, but for averages taking the full set into account provides a much better picture of the full dataframe.  

# 09. Exporting New Dataframe

In [16]:
#exporting updated dataframe as pickle
ords_prods_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'Ords_prod_merged_11262022.pkl'))